MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+Maintenance Job Design Document
+===============================
+
+The (shared) object cache is growing to be large. This includes:
+
+* Number of loose objects.
+* Number of pack-files.
+* Size of pack-files.
+
+There are some things we can do to improve this situation, with varying levels
+of difficulty and benefit. Here are the four items I believe will have a strong
+impact on our data size and performance:
+
+I. Delete loose objects that exist in pack-files.
+II. Place loose objects into pack-files.
+III. Repack some prefetch packs into larger packs.
+IV. Delete packs with no objects referenced by the multi-pack-index.
+
+These are numbered by order of implementation, but they work together in an
+interesting way. If we place loose objects into pack-files (I), then the next
+time we delete loose objects (II) we will delete those objects that we just put
+into a pack-file. If we repack some prefetch packs into larger packs (III), the
+multi-pack-index will use the "newer" pack for all objects in those packs, and
+we can delete them (IV).
+
+In order to perform this maintenance without disrupting users, we can actually
+do these interactions in two independent jobs, with their sub-actions in a
+particular order:
+
+A. Loose Object Maintenance
+  1. Delete loose objects that exist in pack-files.
+  2. Place loose objects into pack-files.
+B. Prefetch Pack Maintenance
+  1. Delete packs with no objects referenced by the multi-pack-index.
+  2. Repack some prefetch packs into larger packs.
+
+Care must be taken to ensure that Git processes that initialized their object
+store before the maintenance   still succeed. Thus, we should never delete
+anything that those processes could be reading from. This is why we delete loose
+objects only if they are referenced by pack-files. This is why we delete
+prefetch packs only if they have no objects referenced by the multi-pack-index.
+Since a second run of these jobs could delete objects that were repacked in the
+previous run, we need to space our job runs such that no Git processes span both
+runs.
+
+Loose Object Job
+--------------
+
+The loose object job can save users who have many loose object downloads. These
+loose objects are populated by several user activities: running `git fetch` will
+get loose commits and trees for the ref tips that are not covered by prefetch
+packs, hydrating a file will trigger a loose blob, and many other. Older
+enlistments have many loose commits and trees due to the behavior of `git fetch`
+before the `--no-show-forced-updates` option.
+
+### Delete loose objects that appear in a pack-file
+
+By running `git prune-packed -q` with the environment variable
+`GIT_OBJECT_DIRECTORY` set to the object cache, Git will delete the loose
+objects that exist in a pack-file. This will use the multi-pack-index to test
+containment, so is only bound by the file-deletion operations.

```

Comment:
If the file is locked and the delete fails does git just skip it an move on, trying again the next time?

Reply:
It uses `unlink_or_warn(path)`, so I believe it is robust against someone having a handle on the file.